### ISSUE TITLE
High memory usage in producer code while using apply_async in celery 4.1 with Redis as a broker

### ISSUE THREAD

**Issue by onkarbhandwalkar3** [celery/celery#4885]
**Update:
O/p of /usr/bin/celery -A proj report** 

software -> celery:4.1.0 (latentcall) kombu:4.1.0 py:3.6.2
            billiard:3.5.0.3 redis:2.10.6
platform -> system:Linux arch:64bit, ELF imp:CPython
loader   -> celery.loaders.app.AppLoader
settings -> transport:redis results:redis://:**@127.0.0.1:6379/1

BROKER_URL: 'redis://:********@127.0.0.1:6379/1'
CELERY_RESULT_BACKEND: 'redis://:********@127.0.0.1:6379/1'
CELERY_INCLUDE: ['services.tasks']
CELERY_TASK_RESULT_EXPIRES: '5'
CELERY_ACCEPT_CONTENT: ['json']
CELERY_RESULT_SERIALIZER: 'json'
CELERY_ACKS_LATE: 'True'
CELERYD_PREFETCH_MULTIPLIER: '1'
BROKER_TRANSPORT_OPTIONS: {
 'visibility_timeout': 43200}

BROKER_TRANSPORT_OPTIONS: {
    'visibility_timeout': 43200}
CELERY_RESULT_SERIALIZER: 'json'
CELERY_ACCEPT_CONTENT: ['json']
CELERY_ACKS_LATE: 'True'
CELERY_TASK_RESULT_EXPIRES: '5'
CELERYD_PREFETCH_MULTIPLIER: '1'

Hi, we have producer code which keeps on running and inserts tasks in a Redis , where consumer code picks tasks from the Redis and perform certain operations. Now after certain days we observe strange behaviour with prouder process where memory usage of producer shoots up drastically. For example: after running 10-15 days memory usage remains ~500 MB but then it shoots up to ~20 GB and then keep on increasing. We observe following logs in producer code:

**Could not insert task with error[maximum recursion depth exceeded]**

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/redis/connection.py", line 590, in send_packed_command
    self._sock.sendall(item)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/redis/client.py", line 2408, in _execute
    return command(*args)
  File "/usr/lib/python3.6/site-packages/redis/connection.py", line 610, in send_command
    self.send_packed_command(self.pack_command(*args))
  File "/usr/lib/python3.6/site-packages/redis/connection.py", line 603, in send_packed_command
    (errno, errmsg))
redis.exceptions.ConnectionError: Error 104 while writing to socket. Connection reset by peer.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "file.py", line 975, in <module>
    serializer=common.serializer_value)
  File "/usr/lib/python3.6/site-packages/celery/app/task.py", line 536, in apply_async
    **options
  File "/usr/lib/python3.6/site-packages/celery/app/base.py", line 736, in send_task
    self.backend.on_task_call(P, task_id)
  File "/usr/lib/python3.6/site-packages/celery/backends/redis.py", line 189, in on_task_call
    self.result_consumer.consume_from(task_id)
  File "/usr/lib/python3.6/site-packages/celery/backends/redis.py", line 76, in consume_from
    self._consume_from(task_id)
  File "/usr/lib/python3.6/site-packages/celery/backends/redis.py", line 82, in _consume_from
    self._pubsub.subscribe(key)
  File "/usr/lib/python3.6/site-packages/redis/client.py", line 2482, in subscribe
    ret_val = self.execute_command('SUBSCRIBE', *iterkeys(new_channels))
  File "/usr/lib/python3.6/site-packages/redis/client.py", line 2404, in execute_command
    self._execute(connection, connection.send_command, *args)
  File "/usr/lib/python3.6/site-packages/redis/client.py", line 2415, in _execute
    connection.connect()
  File "/usr/lib/python3.6/site-packages/redis/connection.py", line 502, in connect
    callback(self)
  File "/usr/lib/python3.6/site-packages/redis/client.py", line 2376, in on_connect
    self.subscribe(**channels)

**Also we get following error message in Redis logs:**

Client id=1665824 addr=127.0.0.1:56318 fd=125 name= age=8 idle=0 flags=N db=1 sub=835676 psub=0 multi=-1 qbuf=0 qbuf-free=32768 obl=0 oll=4259 omem=104778152 events=rw cmd=subscribe scheduled to be closed ASAP for overcoming of output buffer limits.

******Please Note:**  On different servers we get recursion depth exceeded error and error mentioned in Redis.log  but stack trace is not printed on every server. So it seems to suggest that either there is some issue when we exceed recursion depth while performing "apply_async" or issue is with client output buffer of Redis. Also we don't see significant increase in memory usage of Redis as that in case of producer code.**

I am not able to figure out exact root cause, any help will be much appreciated. Thanks in advance.

**Comment by onkarbhandwalkar3** [comment]
Sure, thanks for the response. I will upgrade Redis to the latest release and will report back.

**Comment by onkarbhandwalkar3** [comment]
Hi,
   Strangely we are not able to reproduce this issue on other servers in Dev environment but on production this issue gets reproduced after 15-20 days. So before  jumping towards Redis upgrade,  I am just curious to understand couple of things:

1. Every time we got "recursion depth exceeded error" that resulted in high memory usage of producer. As mentioned in earlier comment , I doubt it  was because of subscription error. Stack trace that I have added in previous comment is because of Redis connection failure. I think that stack trace created some confusion regarding the issue.

2. I am curious to understand situations in which recursive code in apply_async triggers "recursion depth exceeded error".  Is it somehow related to large number of tasks or warning seen in Redis logs 
 ( overcoming of output buffer limits ) ?
